### ISSUE TITLE
Support for `@Inherited` annotation

### ISSUE THREAD

**Issue by trustin** [uber/NullAway#742]
NullAway version: 0.10.9

When a `@Nullable` or `@Nonnull` annotation is annotated with the `@java.lang.annotation.Inherited` meta-annotation, NullAway should respect it. For example, NullAway thinks `Bar.foo()` must return non-null because it is not annotated with `@Nullable`, while it is actually nullable because `Bar` extends `Foo` and `Foo.foo()` is `@Nullable` and `@Nullable` is `@Inherited`. As a result, the following test case fails:

```java
package com.uber.nullaway;

import org.junit.Test;

public class NullAwayInheritedTests extends NullAwayTestsBase {

  @Test
  public void checkNotNullTest() {
    defaultCompilationHelper
        .addSourceLines(
            "Nullable.java",
            "package com.uber;",
            "import java.lang.annotation.ElementType;",
            "import java.lang.annotation.Inherited;",
            "import java.lang.annotation.Retention;",
            "import java.lang.annotation.RetentionPolicy;",
            "import java.lang.annotation.Target;",
            "@Inherited",
            "@Retention(RetentionPolicy.RUNTIME)",
            "@Target({ ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE_USE })",
            "public @interface Nullable {}")
        .addSourceLines(
            "Foo.java",
            "package com.uber;",
            "public class Foo {",
            "  @Nullable Object foo() {",
            "    return null;",
            "  }",
            "}")
        .addSourceLines(
            "Bar.java",
            "package com.uber;",
            "public class Bar extends Foo {",
            "  @Override Object foo() {",
            "    return null;",
            "  }",
            "}")
        .doTest();
  }
}
```

NullAway could collect all inherited annotations from the supertypes and interfaces so the annotations annotated with `@Inherited` annotation works as expected. For example, IntelliJ IDEA's inspections understands `@Inherited` annotation.

One pitfall to think about when implementing this feature is the case where `@Nullable` is masked by `@Nonnull` in a subclass:

```java
public class Foo {
  @Nullable foo() { ... }
}

public class Bar extends Foo {
  // 👉 Nullness overridden!
  @Nonnull @Overrides foo() { ... }
}
```

We'll also need to consider the tricky situation where multiple interfaces don't agree on nullness:

```java
public interface NullableFoo {
  @Nullable foo() { ... }
}

public interface NonnullFoo {
  @Nonnull foo() { ... }
}

public class Bar implements NullableFoo, NonnullFoo {
  // 👉 Is this nullable or not?
  // NullAway could report the conflict or treat as non-null.
  @Overrides foo() { ... } 
}
```

**Comment by trustin** [comment]
Hi @msridhar. Thanks for sharing your ideas. I understand that implementing this feature can incur additional complexity as well as introducing backward incompatibility.

I checked the source code and it seemed to me that the changes could be self-contained in `Nullness.java` (specifically, `hasNullableAnnotation()` and `hasNonNullAnnotation`) with a little bit of utility addition to `NullabilityUtil.java` (e.g. checking if the annotation is `@Inherited`.) I might be thinking too naively, so please feel free to correct me if there's more to consider here.

Regarding backward compatibility, how about introducing a configuration option? We could make this feature opt-in.

I understand there's readability issue with inherited annotations, but the semantics of `@Inherited` is very clear and it's part of Java annotation specification, and thus we should respect it. I also think, with the help from tools like NullAway and IntelliJ inspections, a user can easily figure out whether such a method is nullable or not fairly quickly anyway.

If you don't mind and interested in accepting it, I'd be happy to send a PR. 😄

**Comment by msridhar** [comment]
@trustin we'll have to think this over more.  I'm still leaning against adding this support, but I'd like to hear what @lazaroclapp thinks when he's able to comment.

Another bit of context here is that we are working towards support of the [JSpecify](https://jspecify.dev) specification for Java nullness qualifiers.  Our hope is that with a standardized spec, more and more projects will shift from using bespoke nullness annotations to using what is provided by JSpecify.  JSpecify decided against making their annotations inherited:

https://github.com/jspecify/jspecify/issues/14

Of course, NullAway will continue to support other annotations, but given our current direction, supporting `@Inherited` semantics on annotations doesn't seem to be worth the cost in terms of complexity.

To make the complexity costs more concrete, we already have cases where NullAway is behaving strangely under combinations of our supported options for determining nullability, e.g., https://github.com/uber/NullAway/issues/730.  While implemented `@Inherited` support may be relatively easy, the implications in terms of interactions with other features may not be obvious.

**Comment by trustin** [comment]
@lazaroclapp Thoughts?

**Comment by cpovirk** [comment]
Fun fact that I did not know for years: `@Inherited` is [defined](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/annotation/Inherited.html) to apply only to annotations on _classes_, not even on interfaces, let alone methods or parameters. That said, I'm not surprised to see IntelliJ's behavior, and I think I've seen other tools with similar behavior.

Anyway, if you follow a couple links from @msridhar 's link, you can get to [this JSpecify FAQ entry](https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/edit#heading=h.rsvxffwgolm3) on the subject of a sort of "inheritance" of annotations for overrides of `Object.equals`.

**Comment by lazaroclapp** [comment]
> Fun fact that I did not know for years: `@Inherited` is [defined](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/annotation/Inherited.html) to apply only to annotations on classes, not even on interfaces, let alone methods or parameters.

That's a good catch! Given then that the specification doesn't mandate treating `@Inherited` as applying to overriding methods or their arguments, and the very real argument against having "secret" nullability added to values where there is an easy way to declare their nullability _explicitly_, I am kind of inclined _against_ this feature.

There is some repetition, but from a "nullability is a type extension" perspective, then it makes sense that wherever (return and argument) _types_ need to be repeated/given explicitly, the same should be true about nullability annotations. It is just a lot less confusing that way than:

```
@Override 
public void foo(Object o) {
   ...
   o.toString(); // <- error here, because Super.foo(...) happens to take a `@Nullable Object o`
}
```

**Comment by msridhar** [comment]
In light of the comments here, I am going to close this issue.  Again, @trustin we really appreciate the report.  If you have other pain points or issues with NullAway we'd be happy to hear about them and discuss them.

**Comment by trustin** [comment]
Super interesting to know that. I wonder if it's actually a kind of documentation mistakes 😅 Thanks for your kind responses, @msridhar
